### DISCUSSION TITLE
.venv/lib/python3.11/site-packages/websockets/legacy/server.py:1185: DeprecationWarning: remove second argument of ws_handler   warnings.warn("remove second argument of ws_handler", DeprecationWarning)

### DISCUSSION THREAD

**Discussion by nmoreaud** [Kludex/uvicorn#2476] [Potential Issue] [answered]
Hello
I have the following deprecation warning with `uvicorn==0.31 & websockets==13.0.1`.

```log
.venv/lib/python3.11/site-packages/websockets/legacy/server.py:1185: DeprecationWarning: remove second argument of ws_handler
  warnings.warn("remove second argument of ws_handler", DeprecationWarning)
```

In websockets 13.0 (very recent version), there are breaking changes about `async def handler`
https://websockets.readthedocs.io/en/stable/project/changelog.html#id6

I think uvicorn still defines the websocket handler with the previous signature:
https://github.com/encode/uvicorn/blob/master/uvicorn/protocols/websockets/websockets_impl.py#L227C1-L227C111

**Comment by FeodorFitsner** [comment, marked as answer]
Second it. That deprecation warning is confusing to the users of our framework.

I can do a PR if that's OK?

**Reply by Kludex** [reply]
Sure. Please do not drop the minimum websockets version.

**Reply by FeodorFitsner** [reply]
I can do a PR, but let me first to describe the way I'm going to do a fix to support both <= 10.0 and > 10.0 versions of `websockets`.

So, the warning is coming from [`remove_path_argument(ws_handler)`](https://github.com/python-websockets/websockets/blob/main/src/websockets/legacy/server.py#L1161-L1190) method which is used to "strip" the second argument from the handler.

I think in https://github.com/encode/uvicorn/blob/master/uvicorn/protocols/websockets/websockets_impl.py we can define two ws_handler methods, for example:

```py
    def ws_handler(self, protocol: WebSocketServerProtocol) -> Any:  # type: ignore[override]
        return self.ws_handler_legacy(protocol, None)

    async def ws_handler_legacy(self, protocol: WebSocketServerProtocol, path: str) -> Any:  # type: ignore[override]
        self.handshake_completed_event.set()
        await self.wait_closed()
```

 Then [here](https://github.com/encode/uvicorn/blob/master/uvicorn/protocols/websockets/websockets_impl.py#L104) we could check if `websockets.legacy.server` module has `remove_path_argument` method and then provide ws_handler without a second str argument.

Minimal overhead though it's assumed we "know" there is `remove_path_argument` method. How does it sound like?

**Reply by DAVIDS2405** [reply]
@FeodorFitsner : It looks totally fine to me because even though I have the latest versions of both websockets and uvicorn, that warning still pops up when it shouldn't—especially if you're on version 13.0 or higher. @Kludex , it would be awesome if you could check this out. Not to be annoying, but the warning should only apply to folks using a websockets version lower than 13.0.

**Reply by DAVIDS2405** [reply]
@FeodorFitsner I still recommend you keep an eye on it with recent versions of python as there are problems with identation.

**Comment by nmoreaud** [comment]
There is also:
```log
.venv/lib/python3.11/site-packages/websockets/exceptions.py:122: DeprecationWarning: ConnectionClosed.code is deprecated; use Protocol.close_code or ConnectionClosed.rcvd.code
  warnings.warn(  # deprecated in 13.1
.venv/lib/python3.11/site-packages/websockets/exceptions.py:133: DeprecationWarning: ConnectionClosed.reason is deprecated; use Protocol.close_reason or ConnectionClosed.rcvd.reason
  warnings.warn(  # deprecated in 13.1
```

**Reply by Kludex** [reply]
Where's that used?

**Reply by Kludex** [reply]
You are not using the latest uvicorn version, are you?

**Reply by nmoreaud** [reply]
uvicorn==0.31.0

**Reply by Kludex** [reply]
<img width="1263" alt="Screenshot 2024-10-16 at 11 40 46" src="https://github.com/user-attachments/assets/1cc7279b-9ab7-438b-b434-318e3243e1a0">

**Reply by nmoreaud** [reply]
Ok nice, I'm going to test this version

**Comment by Kludex** [comment]
But also... Just to clarify... The warning you see on the description of this discussion was left on purpose because I wanted to support old `websockets` versions, and don't add complexity to the code source.

**Reply by nmoreaud** [reply]
So, even if I use the latest version of `websockets` and `uvicorn`, this warning will still be emitted?
Because for now I still see it:

```
.venv/lib/python3.11/site-packages/websockets/legacy/server.py:1187: DeprecationWarning: remove second argument of ws_handler
uvicorn==0.32.0
websockets==13.1
```

**Reply by Kludex** [reply]
Yes, that's what I'm saying that it was on purpose.

**Reply by gumantov** [reply]
how can I fix that? I am running in this issue and I was not able to fix it by upgrading websockets, unicorn to the latest versions.

**Reply by gumantov** [reply]
`lib/python3.11/site-packages/websockets/legacy/__init__.py:6:` DeprecationWarning: websockets.legacy is deprecated; see https://websockets.readthedocs.io/en/stable/howto/upgrade.html for upgrade instructions
  warnings.warn(  # deprecated in 14.0`

**Reply by FierceTaipan** [reply]
It helped me to remove the libraries 
pip uninstall websockets     
pip uninstall 'uvicorn[standard]'

and install only these
pip install uvicorn
pip install wsproto

**Reply by xmoner** [reply]
I have a problem too... when I run `flet run --ios` in terminal, I get this error:

`
Traceback (most recent call last):
  File "/Users/lukas/Documents/BILEK-STUDIO/CODE/TOOLS/PYTHON/FLET/Ahoy-project/src/main.py", line 4, in <module>
    import database
  File "/Users/lukas/Documents/BILEK-STUDIO/CODE/TOOLS/PYTHON/FLET/Ahoy-project/src/database.py", line 2, in <module>
    from supabase import create_client, Client
  File "/usr/local/lib/python3.10/site-packages/supabase/__init__.py", line 13, in <module>
    from realtime import AuthorizationError, NotConnectedError
  File "/usr/local/lib/python3.10/site-packages/realtime/__init__.py", line 9, in <module>
    from ._async.client import AsyncRealtimeClient
  File "/usr/local/lib/python3.10/site-packages/realtime/_async/client.py", line 11, in <module>
    import websockets
ModuleNotFoundError: No module named 'websockets'
^C%                                                                                                                                                                                                                                                              lukas@lukass-macbook-pro-1 Ahoy-project % flet run --ios
/usr/local/lib/python3.10/site-packages/websockets/legacy/__init__.py:6: DeprecationWarning: websockets.legacy is deprecated; see https://websockets.readthedocs.io/en/stable/howto/upgrade.html for upgrade instructions
  warnings.warn(  # deprecated in 14.0 - 2024-11-09
/usr/local/lib/python3.10/site-packages/uvicorn/protocols/websockets/websockets_impl.py:17: DeprecationWarning: websockets.server.WebSocketServerProtocol is deprecated
  from websockets.server import WebSocketServerProtocol
App is running on: http://10.0.1.38:8551/src/main.py
`
Anyone could help?

**Reply by DAVIDS2405** [reply]
@xmoner hello bro i read the problem and only is a warning independently if the version of use in websockets i use the lasted version in uvicorn and websockets and I see the warning you can programming normaly bro

**Comment by DeoLeung** [comment]
any news on fixing the warning?

**Comment by nilton-medeiros** [comment]
Dear all,
This issue has been lingering for quite some time — it’s practically celebrating an anniversary. Is there any forecast for a resolution, or are we waiting on external factors?
Identified warning
```Plaintext
.venv\Lib\site-packages\uvicorn\protocols\websockets\websockets_impl.py:17: DeprecationWarning: websockets.server.WebSocketServerProtocol is deprecated
  from websockets.server import WebSocketServerProtocol

D:\PythonWorks\projetos\...\.venv\Lib\site-packages\websockets\legacy\server.py:1178: DeprecationWarning: remove second argument of ws_handler
  warnings.warn("remove second argument of ws_handler", DeprecationWarning)
``` 
Thank you in advance for your attention. Looking forward to your feedback.

**Comment by stevenh** [comment]
Should this be raised as an issue instead of just a discussion?

**Comment by FeodorFitsner** [comment]
The whole point was switching uvicorn to sansio implementation. Right now default protocol is "auto" which selects "websockets" giving deprecation warnings.

Run uvicorn with `websockets-sansio` protocol implementation.

Command line:

```
uvicorn main:app --ws websockets-sansio
```

Code:

```python
  config = uvicorn.Config(
      app, host=host, port=port, log_level=log_level, ws="websockets-sansio"
  )
  server = uvicorn.Server(config)
```

Hope that helps.
